openedx/core/djangoapps/content/learning_sequences/models.py

```diff
@@ -121,6 +125,7 @@ class LearningSequence(TimeStampedModel):
     class Meta:
         unique_together = [
             ['learning_context', 'usage_key'],
+            ['learning_context', 'usage_key_hash'],
```

Comment:
The fact that `usage_key` is unique with the learning_context (instead of unique on its own) concerns me a bit.

Can we count on my new `get_learning_sequence` and `get_learning_sequence_by_hash` functions to unambiguously load a single sequence for any given usage key?

Reply:
FWIW, if you want to make `UsageKey` globally unique, I'm fine with it, given that no Old Mongo stuff is ever going to make it into this model. But I would be worried about hash collisions over the long haul if we said `usage_key_hash` was to be globally unique.

Reply:
@ormsbee
> Maybe make get_learning_sequence_by_hash take a LearningContextKey param as well?

I could certainly do that, but it would involve requiring an additional `learning_context_key` (or `course_key`) parameter to the hashed-key version of the Sequence Metadata API, since we cannot derive the `learning_context_key` from the hashed sequence usage key.

> but not a single value lookup on the usage key hash itself. 

Can you help me understand why we don't enforce uniqueness of `usage_key` in this table? Is there ever a situation where the same UsageKey could be used in multiple learning contexts? I was under the impression that every UsageKey contained a specific LearningContextKey.

Reply:
Whoops, overlapping responses.

> FWIW, if you want to make usage_key globally unique, I'm fine with it, given that no Old Mongo stuff is ever going to make it into this model.

Cool.

> But I would be worried about hash collisions over the long haul if we said usage_key_hash was to be globally unique.

Gotcha. I'm going to do some math and see if I agree.

Reply:
If the math shows that we've got room to spare to make it globally unique accounting for future growth, then we should probably reduce the hash size to give smaller URLs. :-P

Reply:
Eh, nvm that last comment. I do think there's value in making the hard-to-read machine part of the URL as short as possible, but global uniqueness is also a worthy goal. I do want to be careful that we really won't hit collisions down the road though, because URLs are a really long term contract. I had originally thought this was going to be implemented by presenting the mapping in the outline call, and letting the frontend do the mapping back to the usage keys for sequence calls.

Reply:
@ormsbee 

> I do think there's value in making the hard-to-read machine part of the URL as short as possible, but global uniqueness is also a worthy goal. I do want to be careful that we really won't hit collisions down the road though, because URLs are a really long term contract. 

I hear you here.

>  I had originally thought this was going to be implemented by presenting the mapping in the outline call, and letting the frontend do the mapping back to the usage keys for sequence calls.

The issue @KristinAoki found with this approach is that the Course Metadata/Blocks/Outline calls are fired in at the same time as Sequence call; we do not wait for any of the former to complete before triggering the latter. So, the mapping isn't available at the time of the Sequence call, requiring that the Sequence API handle both usage keys and hashed keys.

The three approaches I see are:
1) Merge this PR, declaring that the hashed sequence keys globally unique, so that they can be passed to the Seq Meta API unambiguously.
2) Merge this PR with the change that the Seq Meta API will require a context_key as an additional argument when a hashed sequence key is provided, allowing us to interpret hashed keys only in the scope of a learning context.
3) Scrap this PR. Instead, in the MFE, block on the response of the Outline API call before triggering the Seq Meta API call. 
Make it so the Outline API is the only API that is aware of hashed keys.

Option (3) would make it so that new URL feature is contingent on the Outlines API being enabled for the course.

Do you have a preference amongst those three? Fwiw, until we enable this in production, we do have some flexibility to switch from one approach to the other after merging everything if needed.

Reply:
I can probably be convinced otherwise, but my current preference is for (3), because it keeps the understanding of this mapping isolated to one API. I could see things getting painful in the long run if the short hash starts becoming an alternate identifier that different parts of the system start using, and some APIs understand it and some don't. Following that, probably (1), though I'd want to know how much headroom we have for collisions in that case.

Reply:
FWIW, I am okay with provisionally merging this (1) as part of the rollout process.

Reply:
> I can probably be convinced otherwise, but my current preference is for (3), because it keeps the understanding of this mapping isolated to one API. I could see things getting painful in the long run if the short hash starts becoming an alternate identifier that different parts of the system start using, and some APIs understand it and some don't. 

I agree.

> FWIW, I am okay with provisionally merging this (1) as part of the rollout process.

I will do this, but make a follow-up ticket to move to approach (3) being switching this on in production.

Reply:
@ormsbee I pushed a new commit that:
* Adds a uniqueness constraint for `usage_key` on `LearningSequences`.
* Removes the existing unique-together constraint for  `(usage_key, learning_context_key)`.
* Adds a unique-together constraint for `(usage_key_hash, learning_context_key)`, capturing the state of affairs that we want to get to.
* Adds a comment to `get_learning_sequence_by_hash` explaining that we're _assuming_ for now that the hashes are globally unique, but not committing to it. I noted that the hash-based API should be considered experimental.
* Links to a new placeholder ticket https://openedx.atlassian.net/browse/TNL-8638 in several places. We can use that ticket to define follow-up work.

Reply:
copying from Slack:
> Dave: Only real concern is the size of the unique index creation, which will be something like 700K rows on prod. Which isn’t quite at the level of real worry, but is on the threshold of it.

> Kyle: would you prefer if we leave the usage_key index as a unique_together with learning_context_key for now (still assuming in get_learning_sequence that usage_key is unique, though), and then merge the index change later when we more time to think about it / respond to it if it goes awry?